server/src/main/java/org/opensearch/index/store/remote/file/OnDemandBlockIndexInput.java

```diff
@@ -144,56 +146,56 @@ public long length() {
 
     @Override
     public byte readByte() throws IOException {
-        if (currentBlock == null) {
+        if (blockHolder.block == null) {
```

Comment:
There could be a raise (fe between `readByte()` and `close()`) when `blockHolder.block` becomes `null` after the check, the safer way would be:

```
       IndexInput block = blockHolder.block;
       if (block == null) {
            // seek to the beginning
            seek(0);
        } else if (currentBlockPosition() >= blockSize) {
            int blockId = currentBlockId + 1;
            demandBlock(blockId);
        }
        return block.readByte();
```

Reply:
This class (like org.apache.lucene.store.IndexInput) is not thread safe (I'll add it to the class doc). The cleaning action is technically on a separate thread, but also guaranteed to happen when the instance can no longer be used so there's no chance of it racing with any of these methods. I agree your suggestion is the right way to handle a concurrency issue like this, but probably not necessary here. What do you think?

Reply:
> This class (like org.apache.lucene.store.IndexInput) is not thread safe (I'll add it to the class doc).

Yes, should be enough to clarify, thank you